static/app/views/replays/detail/memoryChart.tsx

```diff
@@ -159,5 +202,86 @@ const MemoryChartWrapper = styled('div')`
   border: 1px solid ${p => p.theme.border};
 `;
 
-const MemoizedMemoryChart = memo(MemoryChart);
-export default MemoizedMemoryChart;
+const MemoizedMemoryChart = memo(
+  forwardRef<ReactEchartsRef, Props>((props, ref) => (
+    <MemoryChart forwardedRef={ref} {...props} />
+  ))
+);
+
+interface MemoryChartContainerProps extends Props {
+  currentHoverTime: number | undefined;
+  currentTime: number;
+}
+
+/**
+ * This container is used to update echarts outside of React. `currentTime` is
+ * the current time of the player -- if replay is currently playing, this will be
+ * updated quite frequently causing the chart to constantly re-render. The
+ * re-renders will conflict with mouse interactions (e.g. hovers and
+ * tooltips).
+ *
+ * We need `MemoryChart` (which wraps an `<AreaChart>`) to re-render as
+ * infrequently as possible, so we use React.memo and only pass in props that
+ * are not frequently updated.
+ * */
+function MemoryChartContainer({
+  currentTime,
+  currentHoverTime,
+  startTimestamp = 0,
+  ...props
+}: MemoryChartContainerProps) {
+  const chart = useRef<ReactEchartsRef>(null);
+  const theme = useTheme();
+
+  useEffect(() => {
+    if (!chart.current) {
+      return;
+    }
+    const echarts = chart.current.getEchartsInstance();
+
+    echarts.setOption({
+      series: [
+        {
+          id: 'currentTime',
+          markLine: {
+            data: [
+              {
+                xAxis: currentTime / 1000 + startTimestamp,
+              },
+            ],
+          },
+        },
+      ],
+    });
+  }, [currentTime, startTimestamp, theme]);
+
+  useEffect(() => {
+    if (!chart.current) {
+      return;
+    }
+    const echarts = chart.current.getEchartsInstance();
+
+    echarts.setOption({
+      series: [
+        {
+          id: 'hoverTime',
+          markLine: {
+            data: [
+              ...(currentHoverTime
+                ? [
+                    {
+                      xAxis: currentHoverTime / 1000 + startTimestamp,
+                    },
+                  ]
+                : []),
+            ],
+          },
+        },
+      ],
+    });
+  }, [currentHoverTime, startTimestamp, theme]);
```

Comment:
Should this be a separate hook @ryan953?

Reply:
Separate hook for the useEffect? 
I would leave it the way it is inside MemoryChartContainer.

Why? It's not super-reusable outside the context of echarts, and only some types of charts at that.

Reply:
Ah I meant vs a single useEffect that calls `setOption` for both series